### PR TITLE
Feature/automatically clear data options

### PR DESCRIPTION
**Pull request by CDRussell**
Task/Issue URL: https://app.asana.com/0/193817002363848/870498839050922
Tech Design URL: 
CC: 

**Description**:
Adds ability for users to configure the app to automatically clear their data.

**When will data be cleared**
This data clearing can be triggered either by killing the app (user-initiated or the OS killing the process), or by the app spending a certain amount of time in the background.

screenshot 2018-12-04 at 16 09 24

**What data will be cleared**
There are two options available for what will be cleared

1. Clear tabs only
1. Clear tabs and all data (all cookies etc...)

screenshot 2018-12-04 at 16 09 35

**Steps to test this PR**:
This PR has a pretty heavy footprint in terms of ways in which it affects the app. As such, I've detailed the likely [test cases ](https://app.asana.com/0/193817002363848/923961231638016).

**New/upgraded libraries**
- [WorkManager](https://developer.android.com/topic/libraries/architecture/workmanager/) introduced to handle scheduling clearing after a certain amount of time.
- [Coroutines](https://kotlinlang.org/docs/reference/coroutines-overview.html) added while restructuring the lower layers to better handling async operations.
- A newer version of [Mockito-Kotlin](https://github.com/nhaarman/mockito-kotlin) included which is needed for testing around coroutines. Note, this comes in a new package so there is a large number of changes to test classes to use the new packages instead of the old ones.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

**Comment by subsymbolic** [comment]
The settings page on API 21 is rendering red text:
settings_api21

File: app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+        @JvmStatic
+        @Parameters(name = "Test case: {index} - {0}")
+        fun testData(): Array<TestCase> {
+
+            val timeNow = System.currentTimeMillis()
+
+            return arrayOf(
+                // no background time recorded - always expected to indicate not enough time passed
+                TestCase(false, APP_EXIT_ONLY, TimeUnit.MINUTES.toMillis(5), timeNow, false),
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: The time **shouldn't** be relevant here, but the API requires the time. So these ensures that we get the expected result regardless of the time when it is `APP_EXIT_ONLY`; ie, the tests are validating that the time isn't relevant to the outcome.

**Comment by CDRussell** [comment]
> The settings page on API 21 is rendering red text:
> settings_api21

Well that's unpleasant! 🙈Good spot.

File: app/src/main/java/com/duckduckgo/app/settings/clear/ClearWhenOption.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+package com.duckduckgo.app.settings.clear
+
+import androidx.annotation.IdRes
+import androidx.annotation.StringRes
+import com.duckduckgo.app.browser.R
+import com.duckduckgo.app.statistics.pixels.Pixel
+import java.util.concurrent.TimeUnit
+
+enum class
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/settings/clear/ClearWhenOption.kt]: I wasn't sure on that at first, but I've just did it to see what it looks like and i think it feels better. Good shout.

File: app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt [hunk 3]
```diff
@@ -60,7 +60,17 @@ interface Pixel {
         SURVEY_CTA_SHOWN(pixelName = "mus_cs"),
         SURVEY_CTA_DISMISSED(pixelName = "mus_cd"),
         SURVEY_CTA_LAUNCHED_SURVEY(pixelName = "mus_cl"),
-        SURVEY_SURVEY_DISMISSED(pixelName = "mus_sd")
+        SURVEY_SURVEY_DISMISSED(pixelName = "mus_sd"),
+
+        AUTOMATIC_CLEAR_DATA_WHAT_OPTION_NONE("macwhat_n"),
+        AUTOMATIC_CLEAR_DATA_WHAT_OPTION_TABS("macwhat_t"),
+        AUTOMATIC_CLEAR_DATA_WHAT_OPTION_TABS_AND_DATA("macwhat_a"),
```
Comment (subsymbolic) [inline app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt]: Could we make this `macwhat_d` using `d` for data?

**Comment by subsymbolic** [comment]
Nice work, tested and working flawlessly across a number of devices:
* Pixel 3 XL, Device, API 28
* Samsung S7 Edge, Device, API 24
* Nexus 7 Tablet, Device, API 22
* Pixel 2, Emulator, API 27
* Nexus 5X, Emulator, API 21

File: app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: I might be missing something but when I look these two `APP_EXIT_ONLY` test cases:
- TestCase(false, APP_EXIT_ONLY, TimeUnit.MINUTES.toMillis(0), timeNow, false)
- TestCase(true, APP_EXIT_ONLY, TimeUnit.MINUTES.toMillis(5), timeNow)

It's not clear to me why they would behave differently depending on whether there is / isn't a background time recored, when the timing is irrelevant. I would expect them to both produce the same result.

[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: Yeah, I see what you're saying.

The check about whether the background time is recorded or not is more of a sanity check. So if we haven't yet recorded a timestamp for when the app is backgrounded, we bail early since the system is not in a known state. So those first few tests marked with `// no background time recorded` are really testing the guard is working.

You could argue that regardless of that, we'd know that for `APP_EXIT_ONLY` that the background time is irrelevant. But it feels stranger to allow that through despite us being in an unknown state; it feels safer in that case to bail out since a value we'd definitely expect to be set hasn't been.

File: app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt (hunk 3 above)
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/statistics/pixels/Pixel.kt]: That makes sense to me 👍 

(FYI, I was using `a` for `all`. I realise now that probably wasn't obvious)

File: app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: I think the second option is preferable out of the two. However, even if the logic of handling `APP_EXIT_ONLY` should be handled elsewhere, and the time keeper class shouldn't receive a `APP_EXIT_ONLY` option, the API permits it (since it's accepting any of the enums values).

Options for what to do if one is passed to the method:

1. throw an exception 
1. check it early and return `false` to bail early
1. don't explicitly check for it at all. in that case, the behaviour will be somewhat undefined.

I'd lean towards 2.

[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: I'm happy to go with 2.

Optional if you think it works and you have time: Accept the time as a param to `hasEnoughTimeElapsed` rather than operating on an enum which may or may not encode time information.

[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: Going with 2, we still need the `clearWhenOption: ClearWhenOption` option passed as a parameter, so i'd rather not also specify the time in addition to that.

[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/androidTest/java/com/duckduckgo/app/fire/DataClearerTimeKeeperTest.kt]: I've pushed some changes; let me know what you think / RR

File: app/src/main/java/com/duckduckgo/app/settings/clear/ClearWhenOption.kt [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ClearWhenOption constructor(
```
[the remark this answers is not in the document]
Comment (CDRussell) [reply app/src/main/java/com/duckduckgo/app/settings/clear/ClearWhenOption.kt]: I like `ClearTimingOption`, but I don't like `ClearDataTypeOption`. I think the latter is less clear than what's there already. 

🤔will ponder better names over an upcoming ☕️ and see if I can come up with alternatives we might both like.

[the remark this answers is not in the document]
Comment (subsymbolic) [reply app/src/main/java/com/duckduckgo/app/settings/clear/ClearWhenOption.kt]: Let's leave it as is. It was optional feedback and it probably just stood as it's a little different but it makes perfect sense.

**Comment by CDRussell** [comment]
> The settings page on API 21 is rendering red text:
> settings_api21

This turned out to be a bigger issue than I first assumed.

The problem is that attributes in a color selector can't be used until API 23. Sadly, there is no tooling support for flagging up that this won't work on older APIs it seems.

I think the RED color is coming from a default somewhere as it can't resolve the actual color to use on Lollipop.

However, the plot thickens a little in that it _does work_ when used in conjunction with a custom view. This is why the `SettingsOptionWithSubtitle` classes work fine and weren't showing up as red. 

As such, I've also create a custom class for the settings with a toggle which just wrap what was already there. This isn't ideal, but if we don't do this, we'll need to avoid using attributes in the color selector which would also make me sad as it's a nice way of handling the different themes' colors.

**TL;DR**
If you object to the custom view approach; let's talk more.
If you don't object, can you do some more testing of the settings screen, particularly on any Lollipop devices you have access to.

**Comment by subsymbolic** [comment]
Wow, that is tricky. I have an alternative for you consider. The new attributed color selector is only needed by `SettingsItem` not `SettingsItem.SettingsSwitch` elements so it makes sense to break the connection between those styles. I suggest we also separate the style and theme components of the `SettingsItem.SettingsSwitch` as they are confusing.

So `SettingsItem.SettingSwitch` becomes just `SettingsSwitch` and we introduce a new `SettingsSwitchTheme` and the style and theme are both applied as appropriate to the switched in `content_settings_general.xml`. This should work across devices.

```
    <style name="SettingsSwitch">
        <item name="android:layout_width">match_parent</item>
        <item name="android:layout_height">wrap_content</item>
        <item name="android:paddingTop">10dp</item>
        <item name="android:paddingStart">@dimen/settingsItemHorizontalPadding</item>
        <item name="android:paddingEnd">@dimen/settingsItemHorizontalPadding</item>
        <item name="android:paddingBottom">10dp</item>
        <item name="android:fontFamily">sans-serif</item>
        <item name="android:textColor">?attr/settingsItemTextColor</item>
        <item name="android:textSize">17sp</item>
        <item name="android:textStyle">normal</item>
        <item name="android:selectable">false</item>
    </style>

    <style name="SettingsSwitchTheme">
        <item name="colorControlActivated">@color/cornflowerBlue</item>
        <item name="colorSwitchThumbNormal">@color/white</item>
        <item name="android:colorForeground">?attr/settingsSwitchBackgroundColor</item>
    </style>
```

**Comment by subsymbolic** [comment]
I'm glad that worked out 😊
Let me know when you would like a rereview and final testing.

**Comment by CDRussell** [comment]
> I'm glad that worked out 😊
> Let me know when you would like a rereview and final testing.

It's ready when you are. Thanks!

**Comment by CDRussell** [comment]
Thanks for the thorough review and all the great suggestions! 👍
